packages/react-core/src/utils/createContextUtilities.tsx

```diff
@@ -0,0 +1,111 @@
+import React from 'react';
+
+import { isUndefined, isString } from '@aws-amplify/ui';
+
+export const INVALID_OPTIONS_MESSAGE =
+  'an `errorMessage` or a `defaultValue` must be provided in `options`';
+
+type OptionsWithDefaultValue<ContextType extends {}> = {
+  contextName: string;
+  defaultValue: ContextType;
+};
+
+type OptionsWithErrorMessage = {
+  contextName: string;
+  errorMessage: string;
+};
+
+type ContextOptions<ContextType extends {}> =
+  | OptionsWithDefaultValue<ContextType>
+  | OptionsWithErrorMessage;
+
+/**
+ * Use a `ContextType` generic and `options` to create:
+ * - `Context`: React Context of type `ContextType`
+ * - `Provider`: React Context `Provider` component exposing the `ContextType`
+ *   as optional props
+ * - `useContext`: Utility Hook exposing the values of `ContextType`
+ *
+ * @template ContextType Type definition of the Context.
+ * > For most use cases the keys of `ContextType` should not be optional in
+ * preference of explicit `undefined` to avoid having optional types on the
+ * Utility Hook return
+ *
+ * @param options Context utility options. Requires a `contextName`, and either
+ * a `defaultValue` of `ContextType` or `errorMessage` allowing for differing
+ * behaviors of the Utility Hook when used outside a parent `Provider`:
+ *
+ * - `defaultValue`: Ensures the Utility Hook returns a default value for
+ *   scenarios **where the missing context values should not impact usage**
+ * - `errorMessage`: Ensures the Utility Hook throws an error for
+ *   scenarios **where the missing context values should prevent** usage
+ *
+ * @returns `Context`, `Provider` Component and `useContext` Utility Hook
+ *
+ * @usage
+ * ```ts
+ * interface StuffContextType {
+ *   things: number;
+ * }
+ *
+ * // with `defaultValue`
+ * const [StuffProvider, useStuff] = createContextUtilities<StuffContextType>({
+ *   contextName: 'Stuff',
+ *   defaultValue: { things: 7 }
+ * });
+ *
+ * // with `errorMessage`
+ * const [StuffProvider, useStuff] = createContextUtilities<StuffContextType>({
+ *   contextName: 'Stuff',
+ *   errorMessage: '`useStuff` must be used in a `StuffProvider`'
+ * });
+ * ```
+ */
+export default function createContextUtilities<ContextType extends {}>(
+  options: ContextOptions<ContextType>
+): {
+  Provider: React.ComponentType<React.PropsWithChildren<Partial<ContextType>>>;
+  useContext: () => ContextType;
+  Context: React.Context<ContextType | undefined>;
+} {
+  const { contextName, defaultValue, errorMessage } =
+    (options as {
+      defaultValue: ContextType;
+      errorMessage: string;
+      contextName: string;
+    }) ?? {};
```

Comment:
Not without some reworking of the TS. Tbh it's probably overkill having it here since TS will throw errors if no param is passed but added it just in case